pkg/cloud/services/gc/loadbalancer.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+)
+
+func (s *Service) deleteLoadBalancers(ctx context.Context, resources []*AWSResource) error {
+	for _, resource := range resources {
+		if !s.isLoadBalancerToDelete(resource) {
+			s.scope.V(4).Info("Resource not a load balancer for deletion", "arn", resource.ARN.String())
+			continue
+		}
+
+		if strings.HasPrefix(resource.ARN.Resource, "loadbalancer/app/") {
+			s.scope.V(5).Info("Deleting ALB for Service", "arn", resource.ARN.String())
+			if err := s.deleteLoadBalancerV2(ctx, resource.ARN.String()); err != nil {
+				return fmt.Errorf("deleting ALB: %w", err)
+			}
+			continue
+		}
+		if strings.HasPrefix(resource.ARN.Resource, "loadbalancer/net/") {
+			s.scope.V(5).Info("Deleting NLB for Service", "arn", resource.ARN.String())
+			if err := s.deleteLoadBalancerV2(ctx, resource.ARN.String()); err != nil {
+				return fmt.Errorf("deleting NLB: %w", err)
+			}
+			continue
+		}
+		if strings.HasPrefix(resource.ARN.Resource, "loadbalancer/") {
+			name := strings.ReplaceAll(resource.ARN.Resource, "loadbalancer/", "")
+			s.scope.V(5).Info("Deleting classic ELB for Service", "arn", resource.ARN.String(), "name", name)
+			if err := s.deleteLoadBalancer(ctx, name); err != nil {
+				return fmt.Errorf("deleting classic ELB: %w", err)
+			}
+			continue
+		}
+	}
+
+	s.scope.V(2).Info("Finished processing tagged resources for load balancers")
+
+	return nil
+}
+
+func (s *Service) deleteTargetGroups(ctx context.Context, resources []*AWSResource) error {
+	for _, resource := range resources {
+		if !s.isTargetGroupToDelete(resource) {
+			s.scope.V(4).Info("Resource not a target group for deletion", "arn", resource.ARN.String())
+			continue
+		}
+
+		name := strings.ReplaceAll(resource.ARN.Resource, "targetgroup/", "")
+		if err := s.deleteTargetGroup(ctx, resource.ARN.String()); err != nil {
+			return fmt.Errorf("deleting target group %s: %w", name, err)
+		}
+	}
+	s.scope.V(2).Info("Finished processing resources for target group deletion")
+
+	return nil
+}
+
+func (s *Service) isLoadBalancerToDelete(resource *AWSResource) bool {
```

Comment:
There is a lot of duplication in all of this doing the same thing slightly differently.

Like, the two delete functions can be one just called with a type.

The delegation is also only a single function with either the arn or the name set.

Reply:
What you could do is create two small functions saying `isLBv1` and `isLBv2` and than inside you can combine the two case checks from above. Replace those with a simple function call and you can actually combine the delegation saying something like:

```go
if isLBv2(arn) { input.LoadBalancerArn = aws.StringlbARN } else { input.LoadBalancerName = aws.String(lbARN) }
```

What do you think? :)))

Reply:
the case above would just be: `case isLBv2(resource.ARN.Resource):`

Reply:
isLoadBalancer and isTargetGroup can be combined:

```go
func (s *Service) isLoadBalancerToDelete(resource *AWSResource) bool {
	return s.checkResource(resource, "loadbalancer")
}

func (s *Service) isTargetGroupToDelete(resource *AWSResource) bool {
	return s.checkResource(resource, "targetgroup")
}


func (s *Service) checkResource(resource *AWSResource, prefix string) bool {
	if resource.ARN.Service != elb.ServiceName {
		s.scope.V(5).Info("Resource is not for "+prefix+" service", "arn", resource.ARN.String(), "check", prefix)
		return false
	}
	if !strings.HasPrefix(resource.ARN.Resource, prefix+"/") {
		s.scope.V(5).Info("Resource is not a load balancer", "arn", resource.ARN.String(), "check", prefix)
		return false
	}
	serviceName := resource.Tags[serviceNameTag]
	if serviceName == "" {
		s.scope.V(5).Info("Resource wasn't created for a Service via CCM", "arn", resource.ARN.String(), "check", prefix)
		return false
	}
	s.scope.V(5).Info("Resource is a load balancer to delete", "arn", resource.ARN.String(), "check", prefix, "servicename", serviceName)

	return true
}

```

Reply:
Changed this along these lines.